app/src/main/java/org/efidroid/efidroidmanager/LokiTool.java

```diff
@@ -0,0 +1,49 @@
+package org.efidroid.efidroidmanager;
+
+import android.util.Log;
+
+import com.stericson.roottools.RootTools;
+
+import static com.android.volley.VolleyLog.TAG;
+
+public class LokiTool {
+    public enum PartitionLabel {
+        BOOT("boot"),
+        RECOVERY("recovery");
+
+        private final String mLabel;
+
+        PartitionLabel(String s) {
+            mLabel = s;
+        }
+
+        public String getLabel() {
+            return mLabel;
+        }
+    }
+
+    static {
+        System.loadLibrary("loki-wrapper");
+    }
+
+    private LokiTool() {
+    }
+
+    private static native int lokiPatch(String partitionLabel, String bootloaderImage, String inImage, String outImage);
+
+    private static native int lokiFlash(String partitionLabel, String image);
+
+    public static boolean patchImage(PartitionLabel label, String bootloaderImage, String inImage, String outImage) {
+        return lokiPatch(label.getLabel(), bootloaderImage, inImage, outImage) == 0;
+    }
+
+    public static boolean flashImage(PartitionLabel label, String image) {
+        // root required for flashing
+        boolean isRootAccess = RootTools.isAccessGiven();
+        if (!isRootAccess) {
+            Log.e(TAG, "flashImage: cannot get root privileges");
```

Comment:
Sry that I missed this previously, but why are you using the volley logtag here?
Also why are you checking for root in first place?
1) it has already been done at this point
2) the call to lokiFlash doesn't need root.